spec/services/mediator_spec.rb

```diff
@@ -0,0 +1,65 @@
+RSpec.describe Mediator do
+  subject { described_class }
+
+  let(:collection) { create(:collection) }
+  let(:user) { create(:user) }
+  let(:file) { nil }
+  let(:filename) { nil }
+  let(:link) { nil }
+  let(:chunking_method) { :basic }
+  let(:chunking_profile) { create(:chunking_profile, method: chunking_method, size: 800) }
+  let(:doc) { create(:document, state: :created, link:, file:, filename:, chunking_profile:, collection:, user:) }
+
+  describe "#ingest document" do
+    context "with web link" do
+      let(:link) { "https://en.wikipedia.org/wiki/Tabloid_journalism" }
+
+      it "fetches web page" do
+        # Catching the message seems to prevent the actual method from being called
+        # So limiting this to example
+        allow(FetchWebDocumentJob).to receive("perform_async")
+
+        subject.ingest(doc)
+
+        expect(FetchWebDocumentJob).to have_received("perform_async").with(doc.id)
+      end
+
+      context "when web page has been fetched" do
+        before do
+          allow(ChunkDocumentJob).to receive("perform_async")
+        end
+
+        it "chunks document" do
+          subject.ingest(doc)
+
+          # Need to drain to cause the job to be run so I can check if the
+          # chunking job was initiated after fetching
+          FetchWebDocumentJob.drain
```

Comment:
Looks like this is testing FetchWebDocumentJob. This makes sense, but should be in that class' spec, not here.

If you avoid testing the other class in this spec, you can throw your `allow` into a `before` block, and this spec will get nice and clean.

Reply:
Oh, did you do this because you want to pass control to the job, and then have that job pass it back to this class?

It's better to avoid depending on another class' behaviour in a spec (with some common exceptions like ActiveRecord classes). This class should have certain behaviour, and if getting called by FetchWebDocumentJob is one of those, just make sure you have a context for it in this spec. I wouldn't rely on the other class to call a second instance of the class you're testing.

Reply:
In the case where `doc` is just a web link we have a flow like this:
- `Mediator#ingest`
- `FetchWebDocumentJob#perform`
- `Mediator#ingest`
- `ChunkDocumentJob#perform`

Normally I would have tested fetch web doc behaviour in it's own test, but in this case it's calling `Mediator#ingest` ... so I put this here. Regardless of where I put it, though, I need `#perform` to be executed to test the right job was invoked.

Reply:
I've added test spec for `FetchWebDocumentJob` to test it's behaviour separately from the `Mediator` spec.